src/test/kotlin/com/exactpro/th2/codec/qfj/QFJCodecTest.kt

```diff
@@ -0,0 +1,393 @@
+    }
+
+    @BeforeAll
+    private fun initMessages() {
+        val fixMessage = quickfix.Message()
+        fixMessage.header.apply {
+            setField(BeginString("FIXT.1.1"))
+            setField(SenderCompID("client"))
+            setField(TargetCompID("server"))
+            setField(MsgType("AE"))
+            addGroup(Group(NoHops.FIELD, HopCompID.FIELD).apply {
+                setField(HopCompID("1"))
+            })
+            addGroup(Group(NoHops.FIELD, HopCompID.FIELD).apply {
+                setField(HopCompID("2"))
+            })
+        }
+
+        fixMessage.apply {
+            setField(ApplID("111"))
+            addGroup(Group(NoSides.FIELD, Side.FIELD).apply {
+                setField(Side('1'))
+                addGroup(Group(NoPartyIDs.FIELD, PartyID.FIELD).apply {
+                    setField(PartyID("party1"))
+                    setField(PartyIDSource('D'))
+                    setField(PartyRole(11))
+                })
+                addGroup(Group(NoPartyIDs.FIELD, PartyID.FIELD).apply {
+                    setField(PartyID("party2"))
+                    setField(PartyIDSource('D'))
+                    setField(PartyRole(56))
+                })
+            })
+
+            addGroup(Group(NoSides.FIELD, Side.FIELD).apply {
+                setField(Side('2'))
+                addGroup(Group(NoPartyIDs.FIELD, PartyID.FIELD).apply {
+                    setField(PartyID("party3"))
+                    setField(PartyIDSource('D'))
+                    setField(PartyRole(11))
+                })
+                addGroup(Group(NoPartyIDs.FIELD, PartyID.FIELD).apply {
+                    setField(PartyID("party4"))
+                    setField(PartyIDSource('D'))
+                    setField(PartyRole(56))
+                })
+            })
+        }
+
+        fixMessage.trailer.apply {
+            setField(SignatureLength(9))
+            setField(Signature("signature"))
+        }
+
+        strFixMessage = fixMessage.toString()
+        val bodyLength: String = getBodyLength(strFixMessage)
+        checksumValue = getChecksum(strFixMessage)
+
+        val bytes = fixMessage.toString().toByteArray()
+        rawMessageGroup = getRawMessageGroup(bytes)
+
+        val fieldsMap: MutableMap<String, Value> = HashMap()
+        fieldsMap[QFJCodec.HEADER] = message().apply {
+            this["BeginString"] = "FIXT.1.1"
+            this["SenderCompID"] = "client"
+            this["TargetCompID"] = "server"
+            this["BodyLength"] = bodyLength
+            this["MsgType"] = "AE"
+            this["NoHops"] = listOf(
+                message().addField("HopCompID", 1),
+                message().addField("HopCompID", 2)
+            )
+        }.toValue()
+        fieldsMap["ApplID"] = "111".toValue()
+        fieldsMap["NoSides"] = ListValue.newBuilder().apply {
+            listOf(
+                addValues(message().apply {
```

Comment:
> listOf

What's it even for? We add return-value of `addValues` (it will be the same builder we're adding values into) into it and then never use it again...